### PR TITLE
Writing fasta files from tree sequences

### PR DESCRIPTION
**Pull request by marianne-aspbury**
I've written some code to cover integrating fasta output from tree sequences into tskit, following from issue https://github.com/tskit-dev/tskit/issues/338 and have included various tests for it.
Happy to hear feedback on this (e.g. @jeromekelleher, @hyanwong).

### CODE REVIEW

**Comment by jeromekelleher** [comment]
Great stuff, thanks @marianne-aspbury. How about we go through this in person tomorrow?

**Comment by marianne-aspbury** [comment]
Yep, sounds good, thanks

File: python/tskit/trees.py [hunk 1]
```diff
@@ -2917,6 +2918,63 @@ def samples(self, population=None, population_id=None):
...
+    def write_fasta(self, output, sequence_ids=None, wrap_width=60):
+        """
+        Writes a fasta formatted file from tree sequence data to the
+        specified file-like object.
+
+        Default `sequence_ids` (i.e. the text immediately following ">") are
+        "tsk_{sample_number}" e.g. "tsk_0", "tsk_1" etc. They can be set by providing
+        a list of strings to the `sequence_ids` argument, which must equal the length
+        of the number of samples. Please ensure that these are unique and compatible with
+        fasta standards, since we do not check this (however, do not include the ">").
```
Comment (jeromekelleher) [inline python/tskit/trees.py]: I think mentioning the ">" is confusing here; just leave it out I'd say.

File: python/tskit/trees.py [hunk 2]
```diff
@@ -2917,6 +2918,63 @@ def samples(self, population=None, population_id=None):
...
+        Example usage:
+
+        .. code-block:: python
+            with open("output.fasta", "w") as fasta_file:
+                ts.write_fasta(fasta_file)
+
+        :param File output: The file-like object to write the fasta output.
+        :param list(str) sequence_ids: A list of string names to uniquely identify
+            each of the sequences in the fasta file. Ids will be printed after ">"
+            so you do not need to include this in the strings.
```
Comment (jeromekelleher) [inline python/tskit/trees.py]: Delete this sentence I think --- mentioning ">" is confusing I think.

File: python/tskit/trees.py [hunk 3]
```diff
@@ -2917,6 +2918,63 @@ def samples(self, population=None, population_id=None):
...
+            accordance with fasta standard outputs. To avoid any line-wrapping of
+            sequences, set `wrap_width = 0`. Otherwise, supply any positive integer.
+        """
+        # if not specified, IDs default to sample index
+        if sequence_ids is None:
+            sequence_ids = ["tsk_{}".format(j) for j in self.samples()]
+        if len(sequence_ids) != self.num_samples:
+            raise ValueError(
+                "sequence_ids must have length equal to the number of samples.")
+        if type(wrap_width) != int:
```
Comment (jeromekelleher) [inline python/tskit/trees.py]: Good idea to check that these are integers, but this is a bit tricky (suppose the input was a numpy integer, e.g.). I'm tempted to let this fall through to ``text_wrap`` --- what happens if we provide a string or double to it?

File: python/tskit/trees.py [hunk 4]
```diff
@@ -2917,6 +2918,63 @@ def samples(self, population=None, population_id=None):
...
+            raise TypeError("wrap_width must be a non-negative integer."
+                            "You may specify `wrap_width=0`"
+                            "if you do not want any wrapping.")
+        if wrap_width < 0:
+            raise ValueError("wrap_width must be a non-negative integer.")
+
+        for j, hap in enumerate(self.haplotypes()):
+            # if user doesn't want any wrapping need alt write method
```
Comment (jeromekelleher) [inline python/tskit/trees.py]: As a matter of principle, it'd be nice to get rid of the duplicated code here. Also, I'm not sure about how well the "\n" s will work cross platform. This is a bit neater/safer (we avoid creating some intermediate strings here)
```
print(">", sequence_ids[j], sep="", file=output)
if wrap_width == 0:
     print(hap, file=output)
else:
    for hap_wrap in textwrap.wrap(hap, wrap_width):
        print(hap_wrap, file=output)
```

File: python/tskit/trees.py (hunk 3 above)
Comment (jeromekelleher) [reply python/tskit/trees.py]: I guess if we want to do it here, the canonical way would be to just convert to int:

```
wrap_width = int(wrap_width)
```
If it's not convertible to int, then it'll throw a type error as we want.

Comment (marianne-aspbury) [reply python/tskit/trees.py]: Okay, fair enough, I just didn't think that the error output was very nice, it'll look like this if I get rid of the integer check.
image

File: python/tskit/trees.py (hunk 4 above)
Comment (marianne-aspbury) [reply python/tskit/trees.py]: Okay, thanks!

File: python/tskit/trees.py (hunk 3 above)
Comment (marianne-aspbury) [reply python/tskit/trees.py]: Ah this comment didn't load before, which approach do you think is better?

> 
> 
> I guess if we want to do it here, the canonical way would be to just convert to int:
> 
> ```
> wrap_width = int(wrap_width)
> ```
> 
> If it's not convertible to int, then it'll throw a type error as we want.

Comment (marianne-aspbury) [reply python/tskit/trees.py]: For now I've just deleted the `raise TypeError()` part of the `write_fasta` function (but it's still in the tests). I can add something back in though

Comment (marianne-aspbury) [reply python/tskit/trees.py]: now added `wrap_width = int(wrap_width)`

**Comment by marianne-aspbury** [comment]
Added functionality for printing fastas on command line, pretty much copied directly from the vcf CLI code and all its tests.

**Comment by hyanwong** [comment]
Sorry - I've not been following this. Note my reservations in https://github.com/tskit-dev/tskit/issues/326#issuecomment-524312997 about making sure that FASTA files (and haplotypes in general) output non-variable sites, at a minimum with a dot, but possibly filling in via a reference genome.

**Comment by petrelharp** [comment]
I don't think we're ready to deal with invariant sites yet, @hyanwong - to do that we need to distinguish tree sequences with integer positions and a reference sequence from the current infinite-sites notions. I'm in favor of doing this, but it shouldn't get in the way of this effort.

**Comment by jeromekelleher** [comment]
> I don't think we're ready to deal with invariant sites yet, @hyanwong - to do that we need to distinguish tree sequences with integer positions and a reference sequence from the current infinite-sites notions. I'm in favor of doing this, but it shouldn't get in the way of this effort.

Agreed. Let's consider what we can do as a stopgap for finite sites as part of #326 (without changing the default behaviour; we don't break people's code). We can then decide what the final semantics of the fasta function should be in #353 (tagged as a 0.2.3 release issue so we don't forget about it).

Let's get this PR finished up and merge first though.

**Comment by hyanwong** [comment]
> I don't think we're ready to deal with invariant sites yet, @hyanwong

We've constructed tree sequences with a site at each integer position, but many sites without mutations. Isn't that the same as an invariant site? But I guess this might be though of as a bit of a hack.

File: python/tskit/trees.py [hunk 5]
```diff
@@ -33,6 +33,7 @@
 import concurrent.futures
 
 import numpy as np
+import textwrap
```
Comment (jeromekelleher) [inline python/tskit/trees.py]: textwrap is part of the standard library, so it belongs in the first group of imports.

File: python/tskit/trees.py [hunk 6]
```diff
@@ -2917,6 +2918,68 @@ def samples(self, population=None, population_id=None):
             samples = samples[sample_population == population]
         return samples
 
+    def write_fasta(self, output, sequence_ids=None, wrap_width=60):
+        """
+        Writes a fasta formatted file from tree sequence data to the
```
Comment (jeromekelleher) [inline python/tskit/trees.py]: This assumes that the user knows what a fasta is. How about something like "Writes haplotype data for samples in FASTA format to the specified file-like object".

File: python/tskit/trees.py [hunk 7]
```diff
@@ -2917,6 +2918,68 @@ def samples(self, population=None, population_id=None):
...
+        # if not specified, IDs default to sample index
+        if sequence_ids is None:
+            sequence_ids = ["tsk_{}".format(j) for j in self.samples()]
+        if len(sequence_ids) != self.num_samples:
+            raise ValueError(
+                "sequence_ids must have length equal to the number of samples.")
+
+        wrap_width = int(wrap_width)
+        if wrap_width < 0:
+            raise ValueError("wrap_width must be a non-negative integer."
```
Comment (jeromekelleher) [inline python/tskit/trees.py]: Need some spaces at the end of these strings here (look at the string that is printed out, the words will be mushed together)

**Comment by marianne-aspbury** [comment]
> 
> 
> Looks great, thanks @marianne-aspbury! Some very minor comments above --- we're ready to merge after these are addressed and the branch is rebased to bring it up to date.

Thanks @jeromekelleher, I've fixed these now

**Pull request merged**